velox/functions/lib/Re2Functions.cpp

```diff
@@ -970,89 +970,267 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
   };
 }
 
-PatternMetadata determinePatternKind(StringView pattern) {
-  vector_size_t patternLength = pattern.size();
-  vector_size_t i = 0;
-  // Index of the first % or _ character.
-  vector_size_t wildcardStart = -1;
-  // Count of wildcard character sequences in pattern.
-  vector_size_t numWildcardSequences = 0;
+std::string unescape(
+    StringView pattern,
+    size_t start,
+    size_t end,
+    std::optional<char> escapeChar) {
+  if (!escapeChar) {
+    return std::string(pattern.data(), start, end - start);
```

Comment:
This needs to be std::string(pattern.data() + start, end - start); Otherwise, lots of ASAN failures.

```
=================================================================
==293875==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x611000e5b4c0 at pc 0x000000319269 bp 0x7ffff14e7e60 sp 0x7ffff14e7620
READ of size 129 at 0x611000e5b4c0 thread T0
SCARINESS: 26 (multi-byte-read-heap-buffer-overflow)
    #0 0x319268 in __interceptor_strlen.part.0 ubsan.c
    #1 0x7f069dbf009d in std::char_traits<char>::length(char const*) fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/char_traits.h:371
    #2 0x7f069dbeff08 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>::basic_string<std::allocator<char>>(char const*, std::allocator<char> const&) fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/basic_string.h:536
    #3 0x7f069ddac65c in facebook::velox::functions::unescape[abi:cxx11](facebook::velox::StringView, unsigned long, unsigned long, std::optional<char>) fbcode/velox/functions/lib/Re2Functions.cpp:979
    #4 0x7f069ddaf13f in facebook::velox::functions::determinePatternKind(facebook::velox::StringView, std::optional<char>) fbcode/velox/functions/lib/Re2Functions.cpp:1203
    #5 0x7f069ddbca05 in facebook::velox::functions::makeLike(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> const&, std::vector<facebook::velox::exec::VectorFunctionArg, std::allocator<facebook::velox::exec::VectorFunctionArg>> const&, facebook::velox::core::QueryConfig const&) fbcode/velox/functions/lib/Re2Functions.cpp:1278
```